File: savo_ws/src/core/savo_head/savo_head/drivers/pca9685_driver.py
# -*- coding: utf-8 -*-

"""Python fallback PCA9685 driver for Robot Savo head diagnostics."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Optional

from savo_head.constants import (
    I2C_BUS_DEFAULT,
    PCA9685_ADDR_DEFAULT,
    PCA9685_PWM_FREQ_HZ_DEFAULT,
    PCA9685_PULSE_MAX_US,
    PCA9685_PULSE_MIN_US,
    PCA9685_SERVO_PERIOD_US,
    PCA9685_TICKS_PER_CYCLE,
)
from savo_head.core.calibration import clamp_pulse_us


MODE1_REG = 0x00
PRESCALE_REG = 0xFE
LED0_ON_L_REG = 0x06
LED0_ON_H_REG = 0x07
LED0_OFF_L_REG = 0x08
LED0_OFF_H_REG = 0x09

MODE1_SLEEP = 0x10
MODE1_RESTART = 0x80

OSCILLATOR_HZ = 25_000_000.0


@dataclass(frozen=True)
class Pca9685Config:
    i2c_bus: int = I2C_BUS_DEFAULT
    address: int = PCA9685_ADDR_DEFAULT
    pwm_frequency_hz: float = PCA9685_PWM_FREQ_HZ_DEFAULT
    dryrun: bool = False

    def normalized(self) -> "Pca9685Config":
        freq = float(self.pwm_frequency_hz)
        if freq <= 0.0:
            freq = PCA9685_PWM_FREQ_HZ_DEFAULT

        return Pca9685Config(
            i2c_bus=int(self.i2c_bus),
            address=int(self.address),
            pwm_frequency_hz=freq,
            dryrun=bool(self.dryrun),
        )


class Pca9685Driver:
    def __init__(self, config: Pca9685Config | None = None, bus: Optional[Any] = None):
        self.config = (config or Pca9685Config()).normalized()
        self._external_bus = bus is not None
        self._bus = bus
        self._opened = False
        self._writes: list[tuple[int, int]] = []

    @property
    def opened(self) -> bool:
        return self._opened

    @property
    def writes(self) -> tuple[tuple[int, int], ...]:
        return tuple(self._writes)

    def open(self) -> None:
        if self._opened:
            return

        if self.config.dryrun:
            self._opened = True
            return

        if self._bus is None:
            try:
                import smbus  # type: ignore
            except ImportError as exc:
                raise RuntimeError("python3-smbus is required for PCA9685 hardware access") from exc

            self._bus = smbus.SMBus(self.config.i2c_bus)

        self.write_byte(MODE1_REG, 0x00)
        self.set_pwm_frequency(self.config.pwm_frequency_hz)
        self._opened = True

    def close(self) -> None:
        if self._bus is not None and not self._external_bus:
            close = getattr(self._bus, "close", None)
            if callable(close):
                close()

        self._bus = None if not self._external_bus else self._bus
        self._opened = False

    def ensure_open(self) -> None:
        if not self._opened:
            self.open()

    def write_byte(self, register: int, value: int) -> None:
        reg = int(register) & 0xFF
        val = int(value) & 0xFF
        self._writes.append((reg, val))

        if self.config.dryrun:
            return

        if self._bus is None:
            raise RuntimeError("PCA9685 bus is not open")

        self._bus.write_byte_data(self.config.address, reg, val)

    def read_byte(self, register: int) -> int:
        reg = int(register) & 0xFF

        if self.config.dryrun:
            return 0x00

        if self._bus is None:
            raise RuntimeError("PCA9685 bus is not open")

        return int(self._bus.read_byte_data(self.config.address, reg))

    def set_pwm_frequency(self, frequency_hz: float) -> int:
        freq = float(frequency_hz)
        if freq <= 0.0:
            raise ValueError("frequency_hz must be positive")

        prescale_value = OSCILLATOR_HZ
        prescale_value /= float(PCA9685_TICKS_PER_CYCLE)
        prescale_value /= freq
        prescale_value -= 1.0
        prescale = int(prescale_value + 0.5)

        old_mode = self.read_byte(MODE1_REG)
        sleep_mode = (old_mode & 0x7F) | MODE1_SLEEP

        self.write_byte(MODE1_REG, sleep_mode)
        self.write_byte(PRESCALE_REG, prescale)
        self.write_byte(MODE1_REG, old_mode)
        self.write_byte(MODE1_REG, old_mode | MODE1_RESTART)

        return prescale

    def set_pwm(self, channel: int, on_ticks: int, off_ticks: int) -> None:
        ch = int(channel)
        if not 0 <= ch <= 15:
            raise ValueError(f"PCA9685 channel must be 0..15, got {channel!r}")

        on = max(0, min(PCA9685_TICKS_PER_CYCLE - 1, int(on_ticks)))
        off = max(0, min(PCA9685_TICKS_PER_CYCLE - 1, int(off_ticks)))

        base = LED0_ON_L_REG + 4 * ch

        self.write_byte(base, on & 0xFF)
        self.write_byte(base + 1, on >> 8)
        self.write_byte(base + 2, off & 0xFF)
        self.write_byte(base + 3, off >> 8)

    def set_servo_pulse_us(self, channel: int, pulse_us: float) -> int:
        pulse = clamp_pulse_us(float(pulse_us))
        ticks = int(pulse * float(PCA9685_TICKS_PER_CYCLE) / float(PCA9685_SERVO_PERIOD_US))
        self.set_pwm(channel, 0, ticks)
        return ticks

    def stop_channel(self, channel: int) -> None:
        self.set_pwm(channel, 0, 0)

    def stop_all(self) -> None:
        for channel in range(16):
            self.stop_channel(channel)


def pulse_us_to_ticks(pulse_us: float) -> int:
    pulse = clamp_pulse_us(float(pulse_us))
    return int(pulse * float(PCA9685_TICKS_PER_CYCLE) / float(PCA9685_SERVO_PERIOD_US))


def validate_servo_pulse_us(pulse_us: float) -> bool:
    value = float(pulse_us)
    return PCA9685_PULSE_MIN_US <= value <= PCA9685_PULSE_MAX_US


__all__ = [
    "MODE1_REG",
    "PRESCALE_REG",
    "LED0_ON_L_REG",
    "LED0_ON_H_REG",
    "LED0_OFF_L_REG",
    "LED0_OFF_H_REG",
    "Pca9685Config",
    "Pca9685Driver",
    "pulse_us_to_ticks",
    "validate_servo_pulse_us",
]

File: savo_ws/src/core/savo_head/src/drivers/pca9685_driver.cpp
#include "savo_head/drivers/pca9685_driver.hpp"

#include <fcntl.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include <cerrno>
#include <cstring>
#include <stdexcept>
#include <utility>

namespace savo_head
{

namespace
{

std::string errno_message(const std::string & prefix)
{
  return prefix + ": " + std::strerror(errno);
}

}  // namespace

Pca9685Driver::Pca9685Driver(Pca9685Config config)
: config_(config.normalized())
{
}

Pca9685Driver::~Pca9685Driver()
{
  close();
}

Pca9685Driver::Pca9685Driver(Pca9685Driver && other) noexcept
{
  *this = std::move(other);
}

Pca9685Driver & Pca9685Driver::operator=(Pca9685Driver && other) noexcept
{
  if (this == &other) {
    return *this;
  }

  close();

  config_ = other.config_;
  fd_ = other.fd_;
  opened_ = other.opened_;
  last_error_ = std::move(other.last_error_);
  writes_ = std::move(other.writes_);

  other.fd_ = -1;
  other.opened_ = false;
  other.last_error_.reset();

  return *this;
}

void Pca9685Driver::open()
{
  if (opened_) {
    return;
  }

  config_ = config_.normalized();
  clear_error();

  if (config_.dryrun) {
    opened_ = true;
    return;
  }

  const auto path = config_.device_path();
  fd_ = ::open(path.c_str(), O_RDWR);

  if (fd_ < 0) {
    set_error(errno_message("failed to open " + path));
    throw std::runtime_error(*last_error_);
  }

  if (::ioctl(fd_, I2C_SLAVE, config_.address) < 0) {
    set_error(errno_message("failed to select PCA9685 I2C address"));
    close();
    throw std::runtime_error(*last_error_);
  }

  opened_ = true;

  try {
    write_byte(kPca9685Mode1Reg, 0x00);
    const auto prescale = set_pwm_frequency(config_.pwm_frequency_hz);
    (void)prescale;
  } catch (...) {
    close();
    throw;
  }
}

void Pca9685Driver::close()
{
  if (fd_ >= 0) {
    ::close(fd_);
    fd_ = -1;
  }

  opened_ = false;
}

bool Pca9685Driver::opened() const
{
  return opened_;
}

bool Pca9685Driver::dryrun() const
{
  return config_.dryrun;
}

int Pca9685Driver::file_descriptor() const
{
  return fd_;
}

const Pca9685Config & Pca9685Driver::config() const
{
  return config_;
}

std::optional<std::string> Pca9685Driver::last_error() const
{
  return last_error_;
}

const std::vector<I2cWrite> & Pca9685Driver::writes() const
{
  return writes_;
}

void Pca9685Driver::clear_error()
{
  last_error_.reset();
}

void Pca9685Driver::clear_writes()
{
  writes_.clear();
}

void Pca9685Driver::write_byte(std::uint8_t reg, std::uint8_t value)
{
  record_write(reg, value);

  if (config_.dryrun) {
    return;
  }

  ensure_open();

  const std::uint8_t data[2] = {reg, value};
  const auto written = ::write(fd_, data, sizeof(data));

  if (written != static_cast<ssize_t>(sizeof(data))) {
    set_error(errno_message("failed to write PCA9685 register"));
    throw std::runtime_error(*last_error_);
  }
}

std::uint8_t Pca9685Driver::read_byte(std::uint8_t reg)
{
  if (config_.dryrun) {
    return 0x00;
  }

  ensure_open();

  const auto reg_write = ::write(fd_, &reg, 1);
  if (reg_write != 1) {
    set_error(errno_message("failed to select PCA9685 register"));
    throw std::runtime_error(*last_error_);
  }

  std::uint8_t value = 0;
  const auto read_count = ::read(fd_, &value, 1);

  if (read_count != 1) {
    set_error(errno_message("failed to read PCA9685 register"));
    throw std::runtime_error(*last_error_);
  }

  return value;
}

int Pca9685Driver::set_pwm_frequency(double frequency_hz)
{
  const auto prescale = pca9685_prescale_for_frequency(frequency_hz);

  const auto old_mode = read_byte(kPca9685Mode1Reg);
  const auto sleep_mode = static_cast<std::uint8_t>((old_mode & 0x7fU) | kPca9685Mode1Sleep);

  write_byte(kPca9685Mode1Reg, sleep_mode);
  write_byte(kPca9685PrescaleReg, static_cast<std::uint8_t>(prescale));
  write_byte(kPca9685Mode1Reg, old_mode);
  write_byte(kPca9685Mode1Reg, static_cast<std::uint8_t>(old_mode | kPca9685Mode1Restart));

  return prescale;
}

void Pca9685Driver::set_pwm(int channel, int on_ticks, int off_ticks)
{
  if (!valid_pca9685_channel(channel)) {
    throw std::out_of_range("PCA9685 channel must be 0..15");
  }

  const auto on = clamp_pca9685_ticks(on_ticks);
  const auto off = clamp_pca9685_ticks(off_ticks);

  const auto base = static_cast<std::uint8_t>(kPca9685Led0OnLReg + 4 * channel);

  write_byte(base, static_cast<std::uint8_t>(on & 0xff));
  write_byte(static_cast<std::uint8_t>(base + 1), static_cast<std::uint8_t>(on >> 8));
  write_byte(static_cast<std::uint8_t>(base + 2), static_cast<std::uint8_t>(off & 0xff));
  write_byte(static_cast<std::uint8_t>(base + 3), static_cast<std::uint8_t>(off >> 8));
}

int Pca9685Driver::set_servo_pulse_us(int channel, double pulse_us)
{
  const auto ticks = pulse_us_to_ticks(pulse_us);
  set_pwm(channel, 0, ticks);
  return ticks;
}

void Pca9685Driver::stop_channel(int channel)
{
  set_pwm(channel, 0, 0);
}

void Pca9685Driver::stop_all()
{
  for (int channel = 0; channel <= 15; ++channel) {
    stop_channel(channel);
  }
}

void Pca9685Driver::ensure_open() const
{
  if (!opened_) {
    throw std::runtime_error("PCA9685 driver is not open");
  }

  if (!config_.dryrun && fd_ < 0) {
    throw std::runtime_error("PCA9685 I2C file descriptor is invalid");
  }
}

void Pca9685Driver::record_write(std::uint8_t reg, std::uint8_t value)
{
  writes_.push_back(I2cWrite{reg, value});
}

void Pca9685Driver::set_error(std::string message)
{
  last_error_ = std::move(message);
}

}  // namespace savo_head
